src/python-bindings/classad2/classad.cpp

```diff
@@ -765,3 +773,141 @@ _classad_unquote( PyObject *, PyObject * args ) {
+        // This was a ClassAdValueError in version 1.
+        PyErr_SetString(PyExc_ValueError, "Unable to flatten expression.");
+        return NULL;
+    }
+    if( out == NULL ) {
+        return convert_classad_value_to_python(value);
+    }
+
+    auto * rv = py_new_classad_exprtree(out);
+    delete out;
+    return rv;
+}
+
+
+static PyObject *
+_classad_external_refs( PyObject *, PyObject * args ) {
+    // _classad_external_refs( self._handle, expr._handle )
+
+    PyObject_Handle * self = NULL;
+    PyObject_Handle * expr = NULL;
+    if(! PyArg_ParseTuple( args, "OO", (PyObject **)& self, (PyObject **)& expr )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * classAd = (ClassAd *)self->t;
+    auto * expression = (ExprTree *)expr->t;
+
+
+    const bool full_names = true;
+    classad::References references;
+    if(! classAd->GetExternalReferences(expression, references, full_names)) {
+        // This was a ClassAdValueError in version 1.
+        PyErr_SetString(PyExc_ValueError, "Unable to determine external references.");
+        return NULL;
+    }
+
+
+    StringList sl;
+    for( const auto & ref : references ) {
+        sl.append(ref.c_str());
+    }
+
+    std::string result = sl.to_string();
+    return PyUnicode_FromString(result.c_str());
+}
+
+
+static PyObject *
+_classad_internal_refs( PyObject *, PyObject * args ) {
+    // _classad_internal_refs( self._handle, expr._handle )
+
+    PyObject_Handle * self = NULL;
+    PyObject_Handle * expr = NULL;
+    if(! PyArg_ParseTuple( args, "OO", (PyObject **)& self, (PyObject **)& expr )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * classAd = (ClassAd *)self->t;
+    auto * expression = (ExprTree *)expr->t;
+
+    const bool full_names = true;
+    classad::References references;
+    if(! classAd->GetInternalReferences(expression, references, full_names)) {
+        // This was a ClassAdValueError in version 1.
+        PyErr_SetString(PyExc_ValueError, "Unable to determine external references.");
+        return NULL;
+    }
+
+
+    StringList sl;
+    for( const auto & ref : references ) {
+        sl.append(ref.c_str());
+    }
+
+    std::string result = sl.to_string();
+    return PyUnicode_FromString(result.c_str());
+}
+
+
+static PyObject *
+_classad_print_json( PyObject *, PyObject * args ) {
+    // _classad_print_json( self._handle )
+
+    PyObject_Handle * handle = NULL;
+    if(! PyArg_ParseTuple( args, "O", (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    std::string str;
+    classad::ClassAdJsonUnParser json_up;
+    json_up.Unparse( str, (ClassAd *)handle->t );
+
+    return PyUnicode_FromString(str.c_str());
+}
+
+
+static PyObject *
+_classad_print_old( PyObject *, PyObject * args ) {
+    // _classad_print_json( self._handle )
```

Comment:
Copy pasta comment mistake here... should be _classad_print_old

Reply:
Just wanted to note this for whenever other changes are added

Reply:
Right-o.